components/canvas/canvas_data.rs

```diff
@@ -160,8 +204,181 @@ impl<'a> PathBuilderRef<'a> {
     }
 }
 
+// TODO(pylbrecht)
+// This defines required methods for DrawTarget of azure and raqote
+// The prototypes are derived from azure's methods.
+pub trait GenericDrawTarget {
+    fn clear_rect(&self, rect: &Rect<f32>);
+    fn copy_surface(&self, surface: SourceSurface, source: Rect<i32>, destination: Point2D<i32>);
+    fn create_gradient_stops(
+        &self,
+        gradient_stops: Vec<GradientStop>,
+        extend_mode: ExtendMode,
+    ) -> GradientStops;
+    fn create_path_builder(&self) -> Box<GenericPathBuilder>;
+    fn create_similar_draw_target(
+        &self,
+        size: &Size2D<i32>,
+        format: SurfaceFormat,
+    ) -> Box<GenericDrawTarget>;
+    fn create_source_surface_from_data(
+        &self,
+        data: &[u8],
+        size: Size2D<i32>,
+        stride: i32,
+    ) -> Option<SourceSurface>;
+    fn draw_surface(
+        &self,
+        surface: SourceSurface,
+        dest: Rect<f64>,
+        source: Rect<f64>,
+        filter: Filter,
+        draw_options: &DrawOptions,
+    );
+    fn draw_surface_with_shadow(
+        &self,
+        surface: SourceSurface,
+        dest: &Point2D<f32>,
+        color: &Color,
+        offset: &Vector2D<f32>,
+        sigma: f32,
+        operator: CompositionOp,
+    );
+    fn fill(&self, path: &Path, pattern: Pattern, draw_options: &DrawOptions);
+    fn fill_rect(&self, rect: &Rect<f32>, pattern: Pattern, draw_options: Option<&DrawOptions>);
+    fn get_format(&self) -> SurfaceFormat;
+    fn get_size(&self) -> Size2D<i32>;
+    fn get_transform(&self) -> Transform2D<f32>;
+    fn pop_clip(&self);
+    fn push_clip(&self, path: &Path);
+    fn set_transform(&self, matrix: &Transform2D<f32>);
+    fn snapshot(&self) -> SourceSurface;
+    fn stroke(
+        &self,
+        path: &Path,
+        pattern: Pattern,
+        stroke_options: &StrokeOptions,
+        draw_options: &DrawOptions,
+    );
+    fn stroke_line(
+        &self,
+        start: Point2D<f32>,
+        end: Point2D<f32>,
+        pattern: Pattern,
+        stroke_options: &StrokeOptions,
+        draw_options: &DrawOptions,
+    );
+    fn stroke_rect(
+        &self,
+        rect: &Rect<f32>,
+        pattern: Pattern,
+        stroke_options: &StrokeOptions,
+        draw_options: &DrawOptions,
+    );
+    fn snapshot_data(&self, f: &Fn(&[u8]) -> Vec<u8>) -> Vec<u8>;
+    fn snapshot_data_owned(&self) -> Vec<u8>;
+}
+
+#[derive(Clone)]
+pub enum ExtendMode {
```

Comment:
Are these enums going to be permanent? I'd imagine that we would have a shared representation with conversions in the backends.

Reply:
I can't decide whether it's worth all the extra duplication this would entail. The answer may become more clear as we start implementing the actual functionality in the raqote backend.

Reply:
Yeah, that's fair, I wasn't entirely sure what the impact of this was, but thought it was worth bringing up